ethdb/olddb/mapmutation.go

```diff
@@ -0,0 +1,304 @@
+		}
+		return value, nil
+	}
+	if m.db != nil {
+		// TODO: simplify when tx can no longer be parent of mutation
+		value, err := m.db.GetOne(table, key)
+		if err != nil {
+			return nil, err
+		}
+
+		return value, nil
+	}
+	return nil, nil
+}
+
+// Can only be called from the worker thread
+func (m *mapmutation) Get(table string, key []byte) ([]byte, error) {
+	value, err := m.GetOne(table, key)
+	if err != nil {
+		return nil, err
+	}
+
+	if value == nil {
+		return nil, ethdb.ErrKeyNotFound
+	}
+
+	return value, nil
+}
+
+func (m *mapmutation) Last(table string) ([]byte, []byte, error) {
+	c, err := m.db.Cursor(table)
+	if err != nil {
+		return nil, nil, err
+	}
+	defer c.Close()
+	return c.Last()
+}
+
+func (m *mapmutation) hasMem(table string, key []byte) bool {
+	m.mu.RLock()
+	defer m.mu.RUnlock()
+	if _, ok := m.puts[table]; !ok {
+		return false
+	}
+
+	_, ok := m.puts[table][string(key)]
+	return ok
+}
+
+func (m *mapmutation) Has(table string, key []byte) (bool, error) {
+	if m.hasMem(table, key) {
+		return true, nil
+	}
+	if m.db != nil {
+		return m.db.Has(table, key)
+	}
+	return false, nil
+}
+
+func (m *mapmutation) Put(table string, key []byte, value []byte) error {
+	m.mu.Lock()
+	defer m.mu.Unlock()
+
+	if _, ok := m.puts[table]; !ok {
+		m.puts[table] = make(map[string][]byte)
+	}
+	var ok bool
+	if _, ok = m.puts[table][string(key)]; !ok {
+		m.size += len(value) - len(m.puts[table][string(key)])
+		m.puts[table][string(key)] = value
+		return nil
+	}
+	m.puts[table][string(key)] = value
+	m.size += len(key) + len(value)
+	m.count++
+	return nil
+}
+
+func (m *mapmutation) Append(table string, key []byte, value []byte) error {
+	return m.Put(table, key, value)
+}
+
+func (m *mapmutation) AppendDup(table string, key []byte, value []byte) error {
+	return m.Put(table, key, value)
+}
+
+func (m *mapmutation) BatchSize() int {
+	m.mu.RLock()
+	defer m.mu.RUnlock()
+	return m.size
+}
+
+func (m *mapmutation) ForEach(bucket string, fromPrefix []byte, walker func(k, v []byte) error) error {
+	m.panicOnEmptyDB()
+	return m.db.ForEach(bucket, fromPrefix, walker)
+}
+
+func (m *mapmutation) ForPrefix(bucket string, prefix []byte, walker func(k, v []byte) error) error {
+	m.panicOnEmptyDB()
+	return m.db.ForPrefix(bucket, prefix, walker)
+}
+
+func (m *mapmutation) ForAmount(bucket string, prefix []byte, amount uint32, walker func(k, v []byte) error) error {
+	m.panicOnEmptyDB()
+	return m.db.ForAmount(bucket, prefix, amount, walker)
+}
+
+func (m *mapmutation) Delete(table string, k, v []byte) error {
+	if v != nil {
+		return m.db.Delete(table, k, v) // TODO: mutation to support DupSort deletes
+	}
+	return m.Put(table, k, nil)
+}
+
+func (m *mapmutation) doCommit(tx kv.RwTx) error {
+	logEvery := time.NewTicker(30 * time.Second)
+	defer logEvery.Stop()
+	count := 0
+	total := float64(m.count)
+
+	for table, bucket := range m.puts {
+		collector := etl.NewCollector("", m.tmpdir, etl.NewSortableBuffer(etl.BufferOptimalSize))
```

Comment:
1. collector.Collect - does CopyBytes inside, don't need to copy here
2. We are sure that data fit in RAM here - any other profit from using Collector here?